kafka-client/src/main/java/dev/responsive/kafka/internal/db/SegmentableKey.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2023 Responsive Computing, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package dev.responsive.kafka.internal.db;
+
+public interface SegmentableKey<T> extends Comparable<T> {
```

Comment:
re: the discussion you were having with @ableegoldman around the awkwardness of the type system - I feel like this interface shouldn't extend `Comparable<T>` and instead we should have:
```java
public class WindowedKey implements SegmentableKey, Comparable<WindowedKey> {
```
This interface doesn't actually require that the `SegmentableKey` is compareable to another `SegmentableKey` (in fact, it doesn't guarantee that at all, so it's a little weird to pass the type parameter through).

If you do this, then you don't need to parameterize the `SegmentParitioner`.